tensorboard/plugins/hparams/summary_v2.py

```diff
@@ -101,6 +172,39 @@ def hparams_config_pb(hparams, metrics, time_created_secs=None):
   )
 
 
+def _normalize_hparams(hparams):
+  """Normalize a dict keyed by `HParam`s and/or raw strings.
+
+  Args:
+    hparams: A `dict` whose keys are `HParam` objects and/or strings
+      representing hyperparameter names, and whose values are
+      hyperparameter values. No two keys may have the same name.
+
+  Returns:
+    A `dict` whose keys are hyperparameter names (as strings) and whose
+    values are the corresponding hyperparameter values.
+
+  Raises:
+    ValueError: If two entries in `hparams` share the same
+      hyperparameter name.
+  """
+  result = {}
+  for (k, v) in six.iteritems(hparams):
+    if isinstance(k, HParam):
+      k = k.name
+    if k in result:
+      raise ValueError("multiple values specified for hparam %r" % (k,))
+    result[k] = v
+  return result
+
+
+def _derive_session_group_name(hparams):
+  # Use `json.dumps` rather than `str` to ensure invariance under string
+  # type (incl. across Python versions) and dict iteration order.
+  jparams = json.dumps(hparams, sort_keys=True, separators=(",", ":"))
+  return hashlib.sha256(jparams.encode("utf-8")).hexdigest()
```

Comment:
We don't really need a full 256 bits of collision avoidance, do we?  N bits gives us about sqrt(N) uses without a collision, so it seems like 128 or fewer bits would be more than enough in our case.

Just thinking that 64 characters is rather long if this If this automatic session name gets exposed in the UI anywhere.  Although I suppose we could always attempt to heuristically detect that it's a hash digest and then abbreviate it.

Reply:
> We don't really need a full 256 bits of collision avoidance, do we? N
> bits gives us about sqrt(N) uses without a collision, so it seems like
> 128 or fewer bits would be more than enough in our case.

$\sqrt(2^N)$, yeah. It’s true that 32 bits suffice for a 99.8% chance of
perfect hashing with up to 4096 hyperparameter configurations, which is
already pushing the performance boundaries of TensorBoard.

> Just thinking that 64 characters is rather long if this If this
> automatic session name gets exposed in the UI anywhere.

It is exposed in the table view, but I have been thinking of removing
that. All the examples that I’ve seen have either a hash or an
incrementing integer their, neither of which is useful. If we remove
that, then we don’t have to have “session groups” or “trials” or
anything in the UI at all.

Note also that these strings in the table view are already truncated
(`text-overflow: ellipsis;`), so truncating from 64- to 32-character
strings in the data wouldn’t actually make an observable difference.
Also, we really shouldn’t shove *either* 32-character hex strings or
64-character hex strings into users’ faces; cutting the length in half
doesn’t actually solve the problem imho. So, given that each session
gropu name appears on disk exactly once across all event files (unlike
tag names, for instance), I’m inclined to keep it as is.

> Although I suppose we could always attempt to heuristically detect
> that it's a hash digest and then abbreviate it.

If we ever want to show these names, this is fine with me.